src/main/scala/is/hail/linalg/BlockMatrix.scala

```diff
@@ -328,20 +327,20 @@ class BlockMatrix(val blocks: RDD[((Int, Int), BDM[Double])],
       1
     }
 
-    hadoop.writeDataFile(uri + metadataRelativePath) { os =>
-      implicit val formats = defaultJSONFormats
-      jackson.Serialization.write(
-        BlockMatrixMetadata(blockSize, nRows, nCols),
-        os)
-    }
-
-    optKeep match {
+    val (partFiles, _) = optKeep match {
       case Some(keep) =>
         blocks.subsetPartitions(keep).writePartitions(uri, writeBlock, Some(keep, partitioner.numPartitions))
       case None =>
         blocks.writePartitions(uri, writeBlock)
     }
 
+    hadoop.writeDataFile(uri + metadataRelativePath) { os =>
```

Comment:
Not necessary in this PR, but you could abstract to `writeMetadata` similar to `readMetadata`. It's used here and in MatrixTable (eventually in RowMatrix). I think it's natural for it to take the GridPartitioner as parameter rather than blockSize, nRows, nCols.

Reply:
Agreed.  Also, before the matrix format becomes non-experimental, we should add the kind of Spec structure that MatrixTable and Table have.